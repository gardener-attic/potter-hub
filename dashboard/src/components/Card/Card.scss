.Card,
.Card__inner {
  height: 100%;
  width: 100%;
}

.Card {
  margin: 0 1% 1.5em;
}

.Card a,
.Card a:hover {
  // TODO: See why we need important here - https://github.com/bitnami/hex/issues/217
  // sass-lint:disable no-important
  text-decoration: none !important;
}

.Card__Icon img {
  max-height: 64px;
}

/* max 4 items per row (default) */
@media screen and (min-width: 480px) {
  .Card.Card-responsive {
    width: 48%;
  }
}

@media screen and (min-width: 768px) {
  .Card.Card-responsive {
    width: 31%;
  }
}

@media screen and (min-width: 1100px) {
  .Card.Card-responsive {
    width: 23%;
  }
}

/* max 3 items per row */
@media screen and (min-width: 768px) {
  .Card.Card-responsive-3 {
    width: 48%;
  }
}

@media screen and (min-width: 1100px) {
  .Card.Card-responsive-3 {
    width: 31%;
  }
}

/* max 2 items per row */
@media screen and (min-width: 768px) {
  .Card.Card-responsive-2 {
    width: 48%;
  }
}
