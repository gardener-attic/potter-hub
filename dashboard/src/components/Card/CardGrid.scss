.CardGrid {
  display: flex;
  margin: 0 -1%;
  justify-content: space-between;
  flex-wrap: wrap;
}

.CardGrid:after {
  /* Display the last row properly: https://stackoverflow.com/a/34816625 */
  flex: auto;
  content: "";
}
