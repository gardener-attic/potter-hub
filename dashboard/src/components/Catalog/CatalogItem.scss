.ListItem__content__info_tag.stable {
  background-color: #1598cb;
}

.ListItem__content__info_tag.incubator {
  background-color: #f58220;
}

.ListItem__content__info_tag_link {
  color: #fff;
  &:hover {
    color: #fff;
  }
  &:visited {
    color: #fff;
  }
}

.ListItem__content__description {
  overflow: hidden;
  height: 80px;
  padding-top: 5px;
  padding-bottom: 5px;
  font-size: 14px;
}
