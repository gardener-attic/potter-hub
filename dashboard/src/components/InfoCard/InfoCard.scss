.ListItem {
  color: #1c2b39;
}

.ListItem__content__title {
  margin: 0;
  width: 100%;
  font-weight: bold;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
  color: #1c2b39;
}

.ListItem__content__info {
  display: flex;
  justify-content: space-between;
}

.ListItem__content__info > p {
  overflow: hidden;
}

.ListItem__content__info_tag {
  float: right;
  margin-left: 3px;
  margin-bottom: 3px;
  color: #fff;
  border-radius: 16px;
  max-width: 150px;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.ListItem__content__info_tag-0 {
  background-color: #1c2b39;
}

.ListItem__content__info_tag-1 {
  background-color: #1598cb;
}

.ListItem__content__info_tag-3 {
  background-color: rgba(238, 107, 11, 0.87);
}

.ListItem__header {
  position: relative;
  display: block;
}

.ListItem__triangle {
  position: absolute;
  top: 0;
  right: 0;
  width: 15%;
}

.ListItem__banner {
  position: absolute;
  bottom: 0;
  background: #008145;
  color: white;
  width: 100%;
  left: 0;
  margin: 0;
  text-align: center;
  padding: 0.25em 0;
  font-size: 14px;
}

.unselectable {
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -khtml-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}
