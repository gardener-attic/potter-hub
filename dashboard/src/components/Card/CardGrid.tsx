import * as React from "react";

import "./CardGrid.css";

export interface ICardGridProps {
  className?: string;
  children?: React.ReactChildren | React.ReactNode | string;
}

interface ICardGridState {
  scrollEndIndex: number
}

const SCROLL_THRESHOLD_TOP_PIXEL = 120;
const SCROLL_THRESHOLD_BOTTOM_TO_LOAD = 0.8;
const ITEMS_PER_SCROLL = 32

class CardGrid extends React.Component<ICardGridProps, ICardGridState> {
  constructor(props: any) {
    super(props);
    this.state = {
      scrollEndIndex: ITEMS_PER_SCROLL,
    };
  }

  public componentDidMount() {
    window.addEventListener("scroll", this.onScroll);
  }

  public componentWillUnmount() {
    window.removeEventListener("scroll", this.onScroll)
  }

  public render() {
    return (
      <div className={`CardGrid padding-v-big ${this.props.className || ""}`}>
        {React.Children.toArray(this.props.children).slice(0, this.state.scrollEndIndex)}
      </div>
    );
  }

  private onScroll = (): void => {
    // calculates, if the current scroll position further than the SCROLL_THRESHOLD_BOTTOM_TO_LOAD threshold
    const isScrollAtBottom = window.innerHeight + document.documentElement.scrollTop >
      document.documentElement.offsetHeight * SCROLL_THRESHOLD_BOTTOM_TO_LOAD;

    // calculates, if the current scroll position is above the SCROLL_THRESHOLD_TOP_PIXEL
    const isScrollAtTop = document.documentElement.scrollTop < SCROLL_THRESHOLD_TOP_PIXEL;

    if (isScrollAtBottom && this.state.scrollEndIndex < React.Children.count(this.props.children)) {
      this.setState((state) => ({
        scrollEndIndex: state.scrollEndIndex + ITEMS_PER_SCROLL
      }))
    }

    if (isScrollAtTop && this.state.scrollEndIndex > ITEMS_PER_SCROLL) {
      this.setState((state) => ({
        scrollEndIndex: ITEMS_PER_SCROLL
      }))
    }
  }
}

export default CardGrid;
