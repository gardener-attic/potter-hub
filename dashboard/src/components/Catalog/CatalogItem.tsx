import * as React from "react";

import placeholder from "../../img/placeholder.png";
import { IChart } from "../../shared/types";
import InfoCard from "../InfoCard";

import "./CatalogItem.css";

interface ICatalogItemProps {
  chart: IChart;
  isFeatured?: boolean;
}

// 3 lines description max
const MAX_DESC_LENGTH = 90;

function trimDescription(desc: string): string {
  if (desc.length > MAX_DESC_LENGTH) {
    // Trim to the last word under the max length
    return desc.substr(0, desc.lastIndexOf(" ", MAX_DESC_LENGTH)).concat("...");
  }
  return desc;
}

const CatalogItem: React.SFC<ICatalogItemProps> = props => {
  const { chart, isFeatured } = props;
  const { icon, name, repo } = chart.attributes;
  const iconSrc = icon ? `api/chartsvc/${icon}` : placeholder;
  const latestChartVersion = chart.relationships.latestChartVersion.data.version
  const latestAppVersion = chart.relationships.latestChartVersion.data.app_version;
  const repoTag = <span className="ListItem__content__info_tag_link">{repo.name}</span>;
  const description = (
    <div className="ListItem__content__description">
      {trimDescription(chart.attributes.description)}
    </div>
  );
  const info = <span>
      Chart:&nbsp;{latestChartVersion}<br/>
      App:&nbsp;{latestAppVersion || "-"}
  </span>
  return (
    <InfoCard
      key={`${repo}/${name}`}
      title={name}
      link={`/charts/${chart.id}`}
      info={info}
      icon={iconSrc}
      description={description}
      tag1Content={repoTag}
      tag1Class={repo.name}
      highlight={isFeatured}
      tagContainerCSS={{ display: "flex", alignItems: "center" }}
    />
  );
};

export default CatalogItem;
