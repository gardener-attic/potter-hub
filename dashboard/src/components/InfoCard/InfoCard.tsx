import * as React from "react";

import { Link } from "react-router-dom";
import placeholder from "../../img/placeholder.png";
import triangle from "../../img/triangle.png";
import Card, { CardContent, CardIcon } from "../Card";
import "./InfoCard.css";

export interface IServiceInstanceCardProps {
  title: string;
  info: string | JSX.Element;
  link?: string;
  icon?: string;
  banner?: string;
  description?: string | JSX.Element;
  tag1Class?: string;
  tag1Content?: string | JSX.Element;
  tag2Class?: string;
  tag2Content?: string | JSX.Element;
  tag3Content?: string | JSX.Element;
  highlight?: boolean;
  tagContainerCSS?: React.CSSProperties | undefined
}

const InfoCard: React.SFC<IServiceInstanceCardProps> = props => {
  const {
    title,
    link,
    info,
    description,
    tag1Content,
    tag1Class,
    tag2Content,
    tag2Class,
    tag3Content,
    banner,
    highlight,
    tagContainerCSS,
  } = props;
  const icon = props.icon ? props.icon : placeholder;

  let tag3: JSX.Element | undefined;

  if (typeof tag3Content === "string") {
    let parsedMetadata;
    try {
      parsedMetadata = JSON.parse(tag3Content);
    } catch {
      // do nothing
    }
    if (parsedMetadata?.bomName) {
      tag3 = (
        <Link to={`/clusterboms/${parsedMetadata.bomName}`}>
          <span
            className={
              "ListItem__content__info_tag type-small type-color-white padding-t-tiny padding-h-normal ListItem__content__info_tag-3"
            }
          >
            BoM: {parsedMetadata.bomName}
          </span>
        </Link>
      );
    } else if (tag3Content.toLocaleLowerCase() === "controllermanaged") {
      tag3 = (
        <span
          className={
            "ListItem__content__info_tag type-small type-color-white padding-t-tiny padding-h-normal ListItem__content__info_tag-3"
          }
        >
          managed by BoM
        </span>
      );
    }
  } else if (tag3Content) {
    tag3 = tag3Content;
  }

  return (
    <Card responsive={true} className="ListItem">
      <Link to={link || "#"} title={title} className="ListItem__header">
        <CardIcon icon={icon} />
        {banner && <p className="ListItem__banner">{banner}</p>}
        {highlight && (
          <img
            src={triangle}
            className="ListItem__triangle"
            title="Featured Chart"
            alt="Featured Chart"
          />
        )}
      </Link>
      <CardContent>
        <div className="ListItem__content">
          <Link to={link || "#"} title={title}>
            <h3 className="ListItem__content__title type-big">{title}</h3>
          </Link>
          {description}
          <div className="ListItem__content__info">
            <p className="margin-reset type-small padding-t-tiny type-color-light-blue">{info}</p>
            <div style={tagContainerCSS}>
              {tag1Content && (
                <span
                  className={`ListItem__content__info_tag ListItem__content__info_tag-1 type-small type-color-white padding-t-tiny padding-h-normal ${tag1Class ||
                    ""}`}
                >
                  {tag1Content}
                </span>
              )}
              {tag2Content && (
                <span
                  className={`ListItem__content__info_tag ListItem__content__info_tag-2 type-small type-color-white padding-t-tiny padding-h-normal ${tag2Class ||
                    ""}`}
                >
                  {tag2Content}
                </span>
              )}
              {tag3}
            </div>
          </div>
        </div>
      </CardContent>
      {props.children}
    </Card>
  );
};

export default InfoCard;
