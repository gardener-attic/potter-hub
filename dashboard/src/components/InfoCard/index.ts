import InfoCard from "./InfoCard";

export default InfoCard;
