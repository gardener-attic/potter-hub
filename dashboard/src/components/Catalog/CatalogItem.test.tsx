import { shallow } from "enzyme";
import context from "jest-plugin-context";
import { cloneDeep } from "lodash";
import * as React from "react";

import { IChart, IRepo } from "../../shared/types";
import { CardIcon } from "../Card";
import InfoCard from "../InfoCard";
import CatalogItem from "./CatalogItem";

jest.mock("../../img/placeholder.png", () => "img/placeholder.png");

const defaultChart = {
  id: "foo",
  attributes: {
    description: "",
    keywords: [""],
    maintainers: [{ name: "" }],
    sources: [""],
    icon: "icon.png",
    name: "foo",
    repo: {} as IRepo,
  },
  relationships: {
    latestChartVersion: {
      data: {
        version: "4.5.6",
        app_version: "1.0.0",
      },
    },
  },
} as IChart;

it("should render an item", () => {
  const wrapper = shallow(<CatalogItem isFeatured={false} chart={defaultChart} />);
  expect(wrapper).toMatchSnapshot();
});

it("should use the default placeholder for the icon if it doesn't exist", () => {
  const chartWithoutIcon = cloneDeep(defaultChart);
  chartWithoutIcon.attributes.icon = undefined;
  const wrapper = shallow(<CatalogItem isFeatured={false} chart={chartWithoutIcon} />);
  // Importing an image returns "undefined"
  expect(
    wrapper
      .find(InfoCard)
      .shallow()
      .find(CardIcon)
      .prop("src"),
  ).toBe(undefined);
});

it("should place a dash if the version is not avaliable", () => {
  const chartWithoutVersion = cloneDeep(defaultChart);
  chartWithoutVersion.relationships.latestChartVersion.data.app_version = "";
  const wrapper = shallow(<CatalogItem isFeatured={false} chart={chartWithoutVersion} />);
  expect(
    wrapper
      .find(InfoCard)
      .shallow()
      .find(".type-color-light-blue")
      .text(),
  ).toBe("Chart: 4.5.6App: -");
});

it("show the chart description", () => {
  const chartWithDescription = cloneDeep(defaultChart);
  chartWithDescription.attributes.description = "This is a description";
  const wrapper = shallow(<CatalogItem isFeatured={false} chart={chartWithDescription} />);
  expect(
    wrapper
      .find(InfoCard)
      .shallow()
      .find(".ListItem__content__description")
      .text(),
  ).toBe(chartWithDescription.attributes.description);
});

context("when the description is too long", () => {
  it("trims the description", () => {
    const chartWithDescription = cloneDeep(defaultChart);
    chartWithDescription.attributes.description =
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum ultrices velit leo, quis pharetra mi vestibulum quis.";
    const wrapper = shallow(<CatalogItem isFeatured={false} chart={chartWithDescription} />);
    expect(
      wrapper
        .find(InfoCard)
        .shallow()
        .find(".ListItem__content__description")
        .text(),
    ).toMatch(/\.\.\.$/);
  });
});
