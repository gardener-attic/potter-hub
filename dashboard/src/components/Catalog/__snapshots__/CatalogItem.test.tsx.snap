// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render an item 1`] = `
<InfoCard
  description={
    <div
      className="ListItem__content__description"
    >
      
    </div>
  }
  highlight={false}
  icon="api/chartsvc/icon.png"
  info={
    <span>
      Chart: 
      4.5.6
      <br />
      App: 
      1.0.0
    </span>
  }
  key="[object Object]/foo"
  link="/charts/foo"
  tag1Content={
    <span
      className="ListItem__content__info_tag_link"
    />
  }
  tagContainerCSS={
    Object {
      "alignItems": "center",
      "display": "flex",
    }
  }
  title="foo"
/>
`;
